File: subsystem_cpp/subsystem/sdk/cwindow.cpp
#include "cwindow.h"
#include "chost_p.h"

struct _self_CWindow {
    CViewController::ref rootViewController;
};

CWindow::CWindow() {
    C_INIT(self);
}

void CWindow::setRootViewController(CViewController::ref controller) {
    self->rootViewController = controller;
}

CViewController::ref CWindow::rootViewController() {
    return self->rootViewController;
}

static void onUILoad(shared_ptr<_self_CWindow> self) {
    if (self->rootViewController) {
        self->rootViewController->viewDidLoad();
    }
}

static void onUIAppear(shared_ptr<_self_CWindow> self) {
    if (self->rootViewController) {
        self->rootViewController->viewDidAppear();
    }
}

static void onUIDisappear(shared_ptr<_self_CWindow> self) {
    if (self->rootViewController) {
        self->rootViewController->viewDidDisappear();
    }
}

void CWindow::makeKeyAndVisible() {

    CHostAddListener(CHostEventUILoad     , bind(onUILoad     , self));
    CHostAddListener(CHostEventUIAppear   , bind(onUIAppear   , self));
    CHostAddListener(CHostEventUIDisappear, bind(onUIDisappear, self));
}

File: subsystem_lua/subsystem/sdk/cwindow.lua
CWindow = extends(CView, {
    rootViewController = nil,
})

function CWindow:init()
    CWindow.super.init(self)
end

function CWindow:setRootViewController(controller)
    self.rootViewController = controller
end

function CWindow:getRootViewController()
    return self.rootViewController
end

function CWindow:makeKeyAndVisible()
    local a = self
    CHost:addListener(CHost.Event.UILoad     , function() onUILoad     (a) end)
    CHost:addListener(CHost.Event.UIAppear   , function() onUIAppear   (a) end)
    CHost:addListener(CHost.Event.UIDisappear, function() onUIDisappear(a) end)
end

function onUILoad(self)
    if self.rootViewController ~= nil then
        self.rootViewController:viewDidLoad()
    end
end

function onUIAppear(self)
    if self.rootViewController ~= nil then
        self.rootViewController:viewDidAppear()
    end
end

function onUIDisappear(self)
    if self.rootViewController ~= nil then
        self.rootViewController:viewDidDisappear()
    end
end
